### PR TITLE
Test with local kinesis

### PR DESCRIPTION
**Pull request by benwaffle**
This PR allows you to test lib-event on your laptop with local versions of kinesis and dynamodb running through https://github.com/localstack/localstack. There is a `localstack.sh` script that will help you spin up localstack in a docker container.

### CODE REVIEW

File: app/io/flow/event/v2/KinesisProducer.scala [hunk 1]
```diff
@@ -143,6 +143,12 @@ case class KinesisProducer[T](
           .withStreamName(config.streamName)
           .withRetentionPeriodHours(72)
       )
+    }.map { _ =>
+      // createStream() immediately returns. we need to wait for the stream to go from CREATING -> ACTIVE.
```
Comment (jlogeart) [inline app/io/flow/event/v2/KinesisProducer.scala]: What happens otherwise? I assume this change solves a concrete issue

**Comment by jlogeart** [comment]
Overall looks promising.

Note I get an exception when running the tests:
```
[warn] application - FlowKinesisError [io.flow.event.v2.KinesisProducer] Stream[DefaultStreamConfig(io.flow.event.v2.AWSCreds@45887247,lib-event-test,development_workstation.lib.event.test.v0.test_event.json,Some(1234),Some(5678),Some(4321),Some(8765),Some(9012),io.flow.lib.event.test.v0.models.TestEvent,io.flow.event.v2.AWSEndpoints@2da83fc3).streamName] could not be created. Error Message: Unable to execute HTTP request: Connect to localhost:4568 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused (Connection refused) {class="io.flow.event.v2.KinesisProducer", stream="development_workstation.lib.event.test.v0.test_event.json"}
com.amazonaws.SdkClientException: Unable to execute HTTP request: Connect to localhost:4568 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused (Connection refused)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleRetryableException(AmazonHttpClient.java:1163)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1109)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:758)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:732)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:714)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:674)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:656)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:520)
	at com.amazonaws.services.kinesis.AmazonKinesisClient.doInvoke(AmazonKinesisClient.java:2803)
	at com.amazonaws.services.kinesis.AmazonKinesisClient.invoke(AmazonKinesisClient.java:2772)
Caused by: org.apache.http.conn.HttpHostConnectException: Connect to localhost:4568 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused (Connection refused)
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:159)
	at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:373)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.amazonaws.http.conn.ClientConnectionManagerFactory$Handler.invoke(ClientConnectionManagerFactory.java:76)
	at com.amazonaws.http.conn.$Proxy18.connect(Unknown Source)
	at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:381)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:237)
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.apache.http.conn.socket.PlainConnectionSocketFactory.connectSocket(PlainConnectionSocketFactory.java:75)
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:142)
	at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:373)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

File: test/io/flow/event/v2/Helpers.scala [hunk 2]
```diff
@@ -44,18 +44,20 @@ trait Helpers {
     eventId
   }
 
-  def consume[T: TypeTag](q: Queue, eventId: String, timeoutSeconds: Int = 35): Record = {
+  def consume[T: TypeTag](q: Queue, eventId: String, timeoutSeconds: Int = 120): Record = {
```
Comment (jlogeart) [inline test/io/flow/event/v2/Helpers.scala]: Why this change?

File: test/io/flow/event/v2/Helpers.scala [hunk 3]
```diff
@@ -44,18 +44,20 @@ trait Helpers {
...
   }
 
-  def consumeUntil[T: TypeTag](q: Queue, eventId: String, timeoutSeconds: Int = 35): Seq[Record] = {
+  def consumeUntil[T: TypeTag](q: Queue, eventId: String, timeoutSeconds: Int = 120): Seq[Record] = {
     val all = scala.collection.mutable.ListBuffer[Record]()
-    // println(s"  --> consumeUntil for eventId[$eventId]")
-    q.consume[T] { rec =>
-      // println(s"  --> record eventId[${rec.eventId}]")
-      all ++= rec
+    println(s"  --> consumeUntil for eventId[$eventId]")
```
Comment (jlogeart) [inline test/io/flow/event/v2/Helpers.scala]: Let's remove the printlns

File: .travis.yml [hunk 4]
```diff
@@ -3,11 +3,16 @@ scala:
...
+services:
+  - docker
 cache:
   directories:
     - $HOME/.ivy2/cache
     - $HOME/.sbt/boot/
+env:
+  - AWS_ACCESS_KEY=x AWS_SECRET_KEY=x AWS_CBOR_DISABLE=true
 script: 
+  - ./localstack.sh
```
Comment (gheine) [inline .travis.yml]: can move this into the before_script section, also why not do `docker run` here directly, like here https://github.com/flowcommerce/experience/blob/master/.travis.yml#L12

File: .travis.yml [hunk 5]
```diff
@@ -3,11 +3,16 @@ scala:
...
 jdk:
   - openjdk10
+services:
+  - docker
 cache:
   directories:
     - $HOME/.ivy2/cache
     - $HOME/.sbt/boot/
+env:
+  - AWS_ACCESS_KEY=x AWS_SECRET_KEY=x AWS_CBOR_DISABLE=true
```
Comment (gheine) [inline .travis.yml]: this is only needed for `sbt test` or anywhere else?

File: app/io/flow/event/v2/KinesisProducer.scala (hunk 1 above)
Comment (benwaffle) [reply app/io/flow/event/v2/KinesisProducer.scala]: This is needed because otherwise the producer might publish events to a stream that is still `CREATING` and not `ACTIVE`. We probably haven't seen this in production because we rarely create new streams; and if we do, I guess we only happen to publish events after the stream has become active.

**Comment by benwaffle** [comment]
@jlogeart 
> Note I get an exception when running the tests

You need to run the local versions of kinesis and dynamodb. Run `./localstack.sh` in the root of the repo. (`docker stop localstack` to stop it)

File: test/io/flow/event/v2/Helpers.scala (hunk 2 above)
Comment (benwaffle) [reply test/io/flow/event/v2/Helpers.scala]: Consumers take a while to start up. 35 seconds is not long enough for the test to pass

File: .travis.yml (hunk 4 above)
Comment (benwaffle) [reply .travis.yml]: :+1: I will move it.

I wanted to add a shell script to run localstack for developing / running tests on your laptop. It is cleaner to use that here than to copy the `docker run` command.

File: .travis.yml (hunk 5 above)
Comment (benwaffle) [reply .travis.yml]: Yes, just needed for testing. `localstack` accepts any key and secret pair. Without these environment variables, the AWS SDK throws an error because it cannot find any credentials (no ec2 role, no `~/.aws/credentials`, ...).

Comment (benwaffle) [reply .travis.yml]: AWS_CBOR_DISABLE is for turning off CBOR support. AWS uses CBOR instead of JSON by default, but localstack doesn't support CBOR yet.

File: app/io/flow/event/v2/AWSConfig.scala [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+)
+
+/**
+  * We run local versions of kinesis and dynamodb for testing through https://github.com/localstack/localstack
+  */
+class AWSEndpoints @Inject() (environment: Environment) {
+  val region = "us-east-1"
+
+  val kinesis = environment.mode match {
+    case Mode.Test => "http://localhost:4568" // localstack
```
Comment (benwaffle) [inline app/io/flow/event/v2/AWSConfig.scala]: I'm concerned that this will require localstack be running for any test, including tests in projects that use lib-event. Do we talk to kinesis in any tests outside lib-event?

File: app/io/flow/event/v2/AWSConfig.scala [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    case _ => s"https://kinesis.$region.amazonaws.com"
```
Comment (jlogeart) [inline app/io/flow/event/v2/AWSConfig.scala]: How about leaving them `None` in `Prod` and keeping it working the way it was before?

**Comment by jlogeart** [comment]
@benwaffle I suggest we add many tests relying on localstack. 

The existing tests rely on mocking the kinesis client (`KinesisProducerSpec`) and mocking the `Queue`.

**Comment by jlogeart** [comment]
@benwaffle Can we run `localstack` from the test themselves?

I suggest introducing a `KinesisIntegrationSpec` that extends `BeforeAndAfterAll` to bring up and down.

**Comment by benwaffle** [comment]
@jlogeart actually, yes. localstack has a library that can do this.

File: app/io/flow/event/v2/AWSConfig.scala (hunk 6 above)
Comment (jlogeart) [reply app/io/flow/event/v2/AWSConfig.scala]: No - they should all rely on mocks, which is perfectly fine. This library should provide mocks for other libraries and services to run their tests.

It is the responsibility of this library to ensure:
 - that it will work in prod (hence `localstack`)
 - mocks are provided and if a library or service tests against them, then it will work on prod

File: app/io/flow/event/v2/AWSConfig.scala (hunk 7 above)
Comment (benwaffle) [reply app/io/flow/event/v2/AWSConfig.scala]: Done - https://github.com/flowcommerce/lib-event/pull/484/commits/ec837a413eeacfc1a43e8d2c18193f020726b72b

File: app/io/flow/event/v2/StreamConfig.scala [hunk 8]
```diff
@@ -100,6 +100,16 @@ object StreamConfig {
         .withMaxLeasesToStealAtOneTime(config.maxLeasesToStealAtOneTime.getOrElse(KinesisClientLibConfiguration.DEFAULT_MAX_LEASES_TO_STEAL_AT_ONE_TIME))
         .withMetricsLevel(MetricsLevel.NONE)
         .withFailoverTimeMillis(30000) // See https://github.com/awslabs/amazon-kinesis-connectors/issues/10
+
+      config.endpoints.kinesis.map { ep =>
```
Comment (jlogeart) [inline app/io/flow/event/v2/StreamConfig.scala]: Note that you may want to use `.foreach` and `.setFoo` instead as this purely side effecting the builder.

File: test/io/flow/event/v2/KinesisProducerSpec.scala [hunk 9]
```diff
@@ -37,58 +50,120 @@ class KinesisProducerSpec extends PlaySpec with MockitoSugar with Inspectors {
...
+      capture.getAllValues.get(1).getRecords must have size 5
+      capture.getAllValues.get(2).getRecords must have size 500
+      capture.getAllValues.get(3).getRecords must have size 100
     }
-    producer.publishBatch(events)
+  }
 
-    val capture: ArgumentCaptor[PutRecordsRequest] = ArgumentCaptor.forClass(classOf[PutRecordsRequest])
-    verify(kinesisClient, times(4)).putRecords(capture.capture())
+  "real KinesisProducer" should {
```
Comment (benwaffle) [inline test/io/flow/event/v2/KinesisProducerSpec.scala]: new test is here. rest of this file is just an indentation change

File: app/io/flow/event/v2/KinesisConsumer.scala [hunk 10]
```diff
@@ -35,16 +34,31 @@ case class KinesisConsumer (
...
+      .withKeyValue("stream", config.streamName)
+      .withKeyValue("worker_id", config.workerId)
 
+  logger_.info("Started")
   exec.execute(worker)
 
   def shutdown(): Unit = {
+    // kill the consumers first
+    logger_.info("Shutting down consumer")
+    if (worker.startGracefulShutdown().get())
```
Comment (jlogeart) [inline app/io/flow/event/v2/KinesisConsumer.scala]: How about adding a timeout here?

Comment (jlogeart) [inline app/io/flow/event/v2/KinesisConsumer.scala]: `.get` may return an exception, in which case it should be caught and logged.

File: test/io/flow/event/v2/KinesisIntegrationSpec.scala [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+import cloud.localstack.Localstack
+import cloud.localstack.ServiceName.{KINESIS, DYNAMO}
+import cloud.localstack.docker.LocalstackDocker
+import cloud.localstack.docker.annotation.LocalstackDockerConfiguration
+import org.scalatest.{BeforeAndAfterAll, Suite}
+import sys.process._
+import scala.language.postfixOps
+import scala.collection.JavaConverters._
+
+trait KinesisIntegrationSpec extends BeforeAndAfterAll { this: Suite =>
```
Comment (jlogeart) [inline test/io/flow/event/v2/KinesisIntegrationSpec.scala]: Nice 👍

File: test/io/flow/event/v2/QueueSpec.scala [hunk 12]
```diff
@@ -96,6 +101,151 @@ TODO: Figure out how to run this test - the worker is not shutting down in time
       q.shutdown
     }
   }
- */
+
+  "produce and consume concurrently" in {
+    withQueue { q =>
```
Comment (jlogeart) [inline test/io/flow/event/v2/QueueSpec.scala]: How about renaming `withQueue` to `withIntegrationQueue`? 
That will help understanding what is using mock vs integration

File: test/io/flow/event/v2/QueueSpec.scala [hunk 13]
```diff
@@ -1,16 +1,26 @@
...
+import org.scalatestplus.play.guice.GuiceOneAppPerSuite
 import play.api.Application
 import play.api.inject.guice.GuiceApplicationBuilder
 
-class QueueSpec extends PlaySpec with GuiceOneAppPerSuite with Helpers {
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success, Try}
+
+class QueueSpec extends PlaySpec with GuiceOneAppPerSuite with Helpers with KinesisIntegrationSpec {
```
Comment (jlogeart) [inline test/io/flow/event/v2/QueueSpec.scala]: Should we rename to `IntegrationQueueSpec` or similar, to mark the difference with `MockQueueSpec`?
